lib/Frontend/ArgsToFrontendOutputsConverter.cpp

```diff
@@ -202,3 +218,225 @@ OutputFilesComputer::deriveOutputFileFromParts(StringRef dir,
+SupplementaryOutputPathsComputer::computeOutputPaths() const {
+  Optional<std::vector<SupplementaryOutputPaths>> pathsFromUser =
+      getSupplementaryOutputPathsFromArguments();
+  if (!pathsFromUser)
+    return None;
+
+  return computeOutputPathsForOneInput(
+      OutputFiles[0], (*pathsFromUser)[0],
+      InputsAndOutputs.firstInputProducingOutput());
+}
+
+Optional<std::vector<SupplementaryOutputPaths>>
+SupplementaryOutputPathsComputer::getSupplementaryOutputPathsFromArguments()
+    const {
+
+  auto objCHeaderOutput = getSupplementaryFilenamesFromArguments(
+      options::OPT_emit_objc_header_path);
+  auto moduleOutput =
+      getSupplementaryFilenamesFromArguments(options::OPT_emit_module_path);
+  auto moduleDocOutput =
+      getSupplementaryFilenamesFromArguments(options::OPT_emit_module_doc_path);
+  auto dependenciesFile = getSupplementaryFilenamesFromArguments(
+      options::OPT_emit_dependencies_path);
+  auto referenceDependenciesFile = getSupplementaryFilenamesFromArguments(
+      options::OPT_emit_reference_dependencies_path);
+  auto serializedDiagnostics = getSupplementaryFilenamesFromArguments(
+      options::OPT_serialize_diagnostics_path);
+  auto loadedModuleTrace = getSupplementaryFilenamesFromArguments(
+      options::OPT_emit_loaded_module_trace_path);
+  auto TBD = getSupplementaryFilenamesFromArguments(options::OPT_emit_tbd_path);
+
+  if (!objCHeaderOutput || !moduleOutput || !moduleDocOutput ||
+      !dependenciesFile || !referenceDependenciesFile ||
+      !serializedDiagnostics || !loadedModuleTrace || !TBD) {
+    return None;
+  }
+  std::vector<SupplementaryOutputPaths> result;
+
+  const unsigned N =
+      InputsAndOutputs.countOfFilesProducingSupplementaryOutput();
+  for (unsigned i = 0; i < N; ++i) {
+    SupplementaryOutputPaths sop;
+    sop.ObjCHeaderOutputPath = (*objCHeaderOutput)[i];
+    sop.ModuleOutputPath = (*moduleOutput)[i];
+    sop.ModuleDocOutputPath = (*moduleDocOutput)[i];
+    sop.DependenciesFilePath = (*dependenciesFile)[i];
+    sop.ReferenceDependenciesFilePath = (*referenceDependenciesFile)[i];
+    sop.SerializedDiagnosticsPath = (*serializedDiagnostics)[i];
+    sop.LoadedModuleTracePath = (*loadedModuleTrace)[i];
+    sop.TBDPath = (*TBD)[i];
+
+    result.push_back(sop);
+  }
+  return result;
+}
+
+// Extend this routine for filelists if/when we have them.
+
+Optional<std::vector<std::string>>
+SupplementaryOutputPathsComputer::getSupplementaryFilenamesFromArguments(
+    options::ID pathID) const {
+  std::vector<std::string> paths = Args.getAllArgValues(pathID);
+
+  const unsigned N =
+      InputsAndOutputs.countOfFilesProducingSupplementaryOutput();
+  if (paths.size() > 0 && paths.size() != N) {
+    Diags.diagnose(SourceLoc(), diag::error_wrong_number_of_arguments,
+                   Args.getLastArg(pathID)->getOption().getName(), N,
+                   paths.size());
+    return None;
+  }
+  while (paths.size() < N)
+    paths.push_back(std::string());
```

Comment:
…which means this can become, unconditionally, [`paths.rezise(N)`](http://en.cppreference.com/w/cpp/container/vector/resize).

Reply:
Thanks for reviewing so quickly! I went down a rathole last night and had to back up.
Maybe have backed up too far. I have rewritten this routine and directly constructs a vector with N blanks if needed.